### PR TITLE
Add nowrap to deflator

### PR DESCRIPTION
**Pull request by tom-walford**
Added support in Deflate for decoding a message overriding the noWrap setting on the Inflator

Not sure what needed to be done to maintain binary compatibility - I added an overloaded method for decodeMessage but added a default parameter to the DeflateDecompressor - hope that's ok?

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by jrudolph** [comment]
Thanks for the contribution, @tom-walford.

The specification is pretty clear that deflate compressed streams need the wrapping. So, I'm wary of adding extra API for non-conforming behavior. On the other hand, it's also good practice to be liberal in what you expect. In the optimal case, the decoder could automatically detect that wrapping and remove it without needing extra API. Could you try if that's possible?

Thanks also the scaladoc changes. Could you split them off into an extra PR that would simplify reviewing and merging.

**Comment by jrudolph** [comment]
Here are some pointers on the wrapping header: https://stackoverflow.com/questions/9050260/what-does-a-zlib-header-look-like. The idea would be to change `DecompressorParsingLogic` in a way that peeks if the first bytes contain the wrapping or not and only then create the `Inflator` with the correct flag.

**Comment by tom-walford** [comment]
Urk, sorry, the docs shouldn't be in this PR!

**Comment by tom-walford** [comment]
@jrudolph Am happy to rework to peek at the head bytes :)

**Comment by tom-walford** [comment]
I've switched it to examining the first 4 bits to check the deflate header - is it possible that this check could pass if we're using no-wrap suitable data and we accidentally use the wrong inflater? I couldn't get much of a clear answer from the libraries themselves

**Comment by jrudolph** [comment]
> I've switched it to examining the first 4 bits to check the deflate header - is it possible that this check could pass if we're using no-wrap suitable data and we accidentally use the wrong inflater?

Hmm, I wonder if that could be case. A deflate block has a 3-*bit* header ([RFC 1951](https://www.ietf.org/rfc/rfc1951.txt)):

```
      3.2.3. Details of block format

         Each block of compressed data begins with 3 header bits
         containing the following data:

            first bit       BFINAL
            next 2 bits     BTYPE

         Note that the header bits do not necessarily begin on a byte
         boundary, since a block does not necessarily occupy an integral
         number of bytes.

         BFINAL is set if and only if this is the last block of the data
         set.

         BTYPE specifies how the data are compressed, as follows:

            00 - no compression
            01 - compressed with fixed Huffman codes
            10 - compressed with dynamic Huffman codes
            11 - reserved (error)
```

If I'm not mistaken the first zlib header byte which is usually 0x78 would be 01111000 which would mean 000 for the 3-bit deflate block header which would be valid as well as a block with no compression. For that it says that:

```
3.2.4. Non-compressed blocks (BTYPE=00)

         Any bits of input up to the next byte boundary are ignored.
         The rest of the block consists of the following information:
```

So, I'd say it's pretty unlikely to see a deflate block starting with 0x78 because a block without compression would probably rather start with 0x00 in the first place. Would be nice to validate that though, as I might have flipped the bits or misread the specs.

**Comment by tom-walford** [comment]
So i ran a few hundred thousand random strings though it (because i'm somewhat paranoid :) ) and they're definitely distinct - notably the CM header always seems to have the first bit set to 1.

If the CM header isn't 8 (the first four bits according to the SO post you linked) then the compression isn't defined, which would match my limited test

**Comment by tom-walford** [comment]
Looking through the spec you're right, but there that risk that if you do have an uncompressed byte string that isn't the final block, then it will attempt to deflate it using nowrap=false.
Granted the code currently checks for 0x08 not 0x78, so I can change that to check for that header if preferred

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Oops ;)

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala [hunk 2]
```diff
@@ -89,10 +89,19 @@ private[http] object DeflateCompressor {
 class DeflateDecompressor(maxBytesPerChunk: Int = Decoder.MaxBytesPerChunkDefault) extends DeflateDecompressorBase(maxBytesPerChunk) {
 
   override def createLogic(attr: Attributes) = new DecompressorParsingLogic {
-    override val inflater: Inflater = new Inflater()
+    var maybeNoWrappedInflater: Option[Inflater] = None
+    override def inflater: Inflater = maybeNoWrappedInflater.getOrElse(new Inflater())
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Hmm, that construction is a bit weird. Maybe just like that:

```scala
var _inflater: Inflater = _
def inflater = if (_inflater ne null) _inflater else throw new IllegalStateException("Inflater used before probing for wrapping")
```

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala [hunk 3]
```diff
@@ -89,10 +89,19 @@ private[http] object DeflateCompressor {
...
 
     override val inflateState = new Inflate(true) {
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Let's create a new state here while probing for the wrapping.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala [hunk 4]
```diff
@@ -89,10 +89,19 @@ private[http] object DeflateCompressor {
...
       override def onTruncation(): Unit = completeStage()
+
+      override def parse(reader: ByteStringParser.ByteReader): ParseResult[ByteString] = {
+        val data = reader.remainingData
+        maybeNoWrappedInflater = maybeNoWrappedInflater
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Here `_inflater` will be initialized for the first time.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala [hunk 5]
```diff
@@ -89,10 +89,19 @@ private[http] object DeflateCompressor {
...
+          .orElse(data.headOption.filter(b ⇒ (b & 0x0F) != 8)
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: We should probably add a `peekByte` method to `ByteReader`. This then needs to be documented with all the information we researched here in this thread.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala [hunk 6]
```diff
@@ -89,10 +89,19 @@ private[http] object DeflateCompressor {
...
+            .fold(Some(new Inflater()))(_ ⇒ Some(new Inflater(true))))
+        super.parse(new ByteStringParser.ByteReader(reader.takeAll()))
+      }
     }
 
     override def afterInflate = inflateState
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Needs to be set to the new initial state.

File: akka-http-tests/src/test/scala/akka/http/scaladsl/coding/DeflateSpec.scala [hunk 7]
```diff
@@ -24,5 +33,20 @@ class DeflateSpec extends CoderSpec {
         ourDecode(ByteString(0, 1, 2, 3, 4))
       }).ultimateCause should be(a[DataFormatException])
     }
+    "properly round-trip encode/decode an HttpRequest using wrapping" in {
+      val request = HttpRequest(POST, entity = HttpEntity(largeText))
```
Comment (jrudolph) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/coding/DeflateSpec.scala]: Would be nice to add a few tests for the edge cases like the no compression deflate compression without wrapping case discussed in the PR.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala (hunk 2 above)
Comment (tom-walford) [reply akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Yeah wasn't sure if you're were ok with nulls - will change it round :)

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala (hunk 1 above)
Comment (tom-walford) [reply akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: yeah... oops :)

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala (hunk 5 above)
Comment (tom-walford) [reply akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala]: Possibly - ByteReader is from Akka Streams though - does that not need a seperate issue/PR?

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by raboof** [comment]
Failure is #1560 - perhaps look at #1565

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by jrudolph** [comment]
Refs #1359.
